Avoid duplicate extensions

CharliePoole opened the pull request:
Fixes #1427 

The initial commit solves the problem of duplicate extension assemlies in the full installation but not when using NuGet packages.

NOTE: We will need to decide on a versioning scheme for extensions. In the current build, extensions all remain as 1.0.0.0.

rprouse commented on src/NUnitEngine/nunit.engine/Services/ExtensionAssembly.cs:
```diff
@@ -0,0 +1,125 @@
+        {
+            get
+            {
+                if (_assemblyDefinition == null)
```
We may not need it here, but should we get in the habit of using Lazy<T> for all lazy initialization? It may not be needed now, but it is easy to implement and future proofs the code.

CharliePoole replied:
That would mean adding the NUnit.System.Linq package with the engine. We planned on that but haven't done it yet. If you think we should, I'll do a separate PR to make sure it doesn't break anything and then merge it .

rprouse replied:
I'm not sure, it was just an observation, do you think it is worth the effort? I would like to see the Linq package added to the engine eventually, but wasn't planning on doing it so soon. Do you think it is worth it?

CharliePoole replied:
I'd rather get the duplicate extension problem solved for the release. Use of Lazy etc. only impacts our dev work.

CharliePoole commented:
This isn't really helping us now, but it will in the future. I added a named property `EngineVersion` to `ExtensionAttribute` in order to indicate the minimum engine version required for an extension. Teamcity extension is set to 3.4, while the others don't use it. I tested by increasing the value and watching in debug to see that the extension was ignored.

CharliePoole commented:
This is ready to merge. In addition to the CI tests passing, I have  tested the ConsoleRunner NuGet package with two different versions of the NUnitProjectLoader package. Only the latest one gets loaded.

In addition to the base fix, I added the EngineVersion named property as described above and updated the AssemblyVersion of our existing extensions to 3.4, matching the nuget packages. The new teamcity listener is versioned as 1.0.0.0 for the assembly and 1.0.0 for the package.

rprouse commented on build.cake:
```diff
@@ -707,7 +707,8 @@ Task("PackageNuGet")
         {
-            Version = packageVersion,
+		    // The teamcity-event-listener extension uses its own versioning
+            Version = "1.0.0" + dbgSuffix,
```
If we are going to be versioning each plugin on it's own update cycle, we should switch each of them to use their own version/AssemblyInfo and extract the version to use here from that. I just did that in a private project with lots of plugins and it worked well. It meant that I only have to update the version in one place. We might even want to do the same for the framework and the engine, pull the common version, rather than update it there and here. Just thinking out loud :smile:
